std/net/isemail.d

```diff
@@ -1863,3 +1857,95 @@ const(T)[] get (T) (const(T)[] str, size_t index, dchar c)
     assert("abc".get(1, 'b') == "b");
     assert("löv".get(1, 'ö') == "ö");
 }
+
+/+
+Replacement for:
+---
+static fourChars = ctRegex!(`^[0-9A-Fa-f]{0,4}$`.to!(const(Char)[]));
+...
+a => a.matchFirst(fourChars).empty
+---
++/
+private static bool isUpToFourHexChars(Char)(scope const(Char)[] s)
+{
+    import std.ascii : isHexDigit;
+    if (s.length > 4) return false;
+    foreach (c; s)
+        if (!isHexDigit(c)) return false;
+    return true;
+}
+
+@nogc nothrow pure @safe unittest
+{
+    assert(!isUpToFourHexChars("12345"));
+    assert(!isUpToFourHexChars("defg"));
+    assert(isUpToFourHexChars("1A0a"));
+}
+
+/+
+Replacement for:
+---
+static ipRegex = ctRegex!(`\b(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}`~
+                    `(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$`.to!(const(Char)[]));
+...
+auto matchesIp = addressLiteral.matchAll(ipRegex).map!(a => a.hit).array;
+----
+Note that only the first item of "matchAll" was ever used in practice
+so we can return `const(Char)[]` instead of `const(Char)[][]` using a
+zero-length string to indicate no match.
++/
+private static const(Char)[] matchIPSuffix(Char)(return const Char[] s)
+{
+    if (s.length < "0.0.0.0".length) return null;
+    size_t end = s.length;
+    // Check the first three `[.]\d{1,3}`
+    foreach (_; 0 .. 3)
+    {
+        size_t start = void;
+        if (end >= 2 && s[end-2] == '.')
+            start = end - 2;
+        else if (end >= 3 && s[end-3] == '.')
+            start = end - 3;
+        else if (end >= 4 && s[end-4] == '.')
+            start = end - 4;
+        else
+            return null;
+        uint x = 0;
+        foreach (i; start + 1 .. end)
+        {
+            uint c = cast(uint) s[i] - '0';
+            if (c > 9) return null;
+            x = x * 10 + c;
+        }
```

Comment:
```d
uint x = s[start + 1 .. end].byCodeUnit.take(2).to!uint.ifThrown(256);
```

Reply:
not `@nogc` and not `nothrow`

Reply:
FYI: `@nogc` isn't a requirement for Phobos (we prefer nice, readable code instead of mir). In this case it will be fixed once the already merged `-dip1008` gets activated by default.
Also the previous `matchAll` isn't neither `@nogc` nor `nothrow` either (and neither is the code which uses it).

Regarding `nothrow`, how about `scope(failure) return null;`?
That would save even more lines.

Reply:
While making things `@nogc nothrow` is nice when ever possible, in this specific instance `isEmail` is not `@nogc nothrow`, so this change will have no effect on the user and there's no benefit.

Reply:
What about converting to `ubyte`? then let `to` do the work of making sure the octet is in range.

FWIW, I like the idea of tagging internal functions `@nogc` and `nothrow`, even if the users aren't, because it helps keep GC usage down. Otherwise, little fixes creep in here and there.

Reply:
>in this specific instance `isEmail` is not `@nogc nothrow`

It probably should be but I'll leave that for another pull request. It looks like `isEmail` is incorrectly using Exceptions rather than assert() to debug logic errors in the function itself ([see forum discussion](https://forum.dlang.org/thread/m07gf1$18jl$1@digitalmars.com)).